### PR TITLE
Add NIP04 support

### PR DESCRIPTION
**Pull request by ntheden**

### CODE REVIEW

**Comment by ethicnology** [comment]
@no-prob 

Please read: https://github.com/ethicnology/dart-nostr/pull/23#pullrequestreview-1347384099

**Comment by ntheden** [comment]
> @no-prob
> 
> Please read: [#23 (review)](https://github.com/ethicnology/dart-nostr/pull/23#pullrequestreview-1347384099)

Yeah, I saw that after creating this PR. This PR has unit tests and integrates with Event class. But unit tests could be expanded and reducing dependencies down to just pointycastle is a good  suggestion. Also I could expand the documentation. I will make those changes a little later if someone else doesn't first, right now working on the project that depends on the nip 4 feature.

File: lib/src/ok.dart
Comment (ethicnology) [inline lib/src/ok.dart:1]: Ok but corresponding unit test needed to keep a high level of code coverage

File: lib/src/eose.dart
Comment (ethicnology) [inline lib/src/eose.dart:1]: Ok but unit tests should be created to keep a high level of code coverage

File: lib/src/event.dart
Comment (ethicnology) [inline lib/src/event.dart:1]: Could you remove all the changes you did to event.dart and create `src/nips/nip_004.dart` ?

In this file you will create a Nip4 which extends Events and add the attributes and methods you want implemented

see the design talk about nip12: https://github.com/ethicnology/dart-nostr/pull/24#issuecomment-1493204442

File: lib/src/kepler.dart
Comment (ethicnology) [inline lib/src/kepler.dart:1]: May you put this into a folder `src/crypto/kepler.dart`

and add as a comment the link to the kepler file source

File: lib/src/operator.dart
Comment (ethicnology) [inline lib/src/operator.dart:1]: May you put this into a folder `src/crypto/operator.dart`

and add as a comment the link to the source file

File: lib/src/settings.dart
Comment (ethicnology) [inline lib/src/settings.dart:1]: I don't understand why you need these variables

File: test/dm_test.dart
Comment (ethicnology) [inline test/dm_test.dart:1]: when you will have created `src/nips/nip_004.dart` you could put this test file in `test/nips/nip_004_test.dart`

**Review by ethicnology — CHANGES_REQUESTED** [review]
Thanks for your work, i add some change you can do in order to keep the library modularity and make it easier to maintain over the time

File: lib/src/eose.dart
Comment (ntheden) [reply lib/src/eose.dart:1]: Agree, also I will move to a separate PR.

File: lib/src/event.dart
Comment (ntheden) [reply lib/src/event.dart:1]: Ok, that is done, however there are still some references to nip4 in `event.dart` which are needed for the receive/deserialization path.

File: lib/src/settings.dart
Comment (ntheden) [reply lib/src/settings.dart:1]: `userPrivateKey` is needed for decryption of DM's. It should be in a configuration file or environment variable, not in a source file. However this library will be used by both mobile and desktop, so a configuration file would not be compatible everywhere. Easy solution until it is decided is to have a `settings.dart`.

**Comment by ethicnology** [comment]
tell me when you think you are done

**Comment by ntheden** [comment]
> tell me when you think you are done

I am done with this current PR.

File: lib/src/event.dart
Comment (ethicnology) [reply lib/src/event.dart:1]: I will rework this part

**Review by ethicnology — APPROVED** [review]
Thank you for your contribution

I will rework some parts to keep it as close as possible as the way the library is already implemented

**Pull request merged**